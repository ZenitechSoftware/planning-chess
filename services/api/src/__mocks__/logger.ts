const mockLogger = {
  info: jest.fn(),
  error: jest.fn(),
  debug: jest.fn(),
};

export default mockLogger;
