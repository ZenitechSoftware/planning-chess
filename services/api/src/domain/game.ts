export interface Turn {
  row?: number;
  tile?: number;
  figure?: string;
  playerId: string;
  score?: number;
  turnType: TurnType;
}

export enum TurnType {
  MoveSkipped = 'MoveSkipped',
  FigurePlaced = 'FigurePlaced',
}
