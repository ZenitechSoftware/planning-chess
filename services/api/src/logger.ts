import winston from 'winston';

const logger = winston.createLogger({
  transports: [
    new winston.transports.Console({
      level: process.env.NODE_ENV === 'test' ? 'crit' : 'debug',
      format: winston.format.combine(
        winston.format.colorize(),
        winston.format.simple(),
      ),
    }),
  ],
});

export default logger;
