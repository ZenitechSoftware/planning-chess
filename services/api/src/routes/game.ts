import express from 'express';
import Joi from 'joi';
import { config } from 'dotenv';

const router = express.Router();

config();

const validationOfPath = () => {
  const uuidPath = `${process.env.GAME_ROOM_UUID}`;
  return Joi.string().valid(uuidPath);
};

router.use((_req, res, next) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  next();
});

router.get('/:id', async (req, res) => {
  const passedUUID = req.params.id;
  const pathValidationSchema = validationOfPath();

  await pathValidationSchema
    .validateAsync(passedUUID)
    .then(() => res.status(200).send({ message: 'Validation Successful' }))
    .catch(() => res.status(400).send({ message: 'Validation Error' }));
});

export default router;
