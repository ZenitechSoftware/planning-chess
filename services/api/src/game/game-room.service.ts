import WebSocket, { WebSocketServer } from 'ws';
import { v4 as uuidv4 } from 'uuid';
import logger from '../logger';
import { Player } from '../domain';
import { PlaceFigureMessage } from '../domain/messages';

interface GameRoom {
  roomId: string;
  server: WebSocketServer;
  players: Map<WebSocket, Player>;
  turns: PlaceFigureMessage[];
}

const rooms = new Map<string, GameRoom>();

export const getOrCreateRoom = (id?: string): GameRoom => {
  const roomId = id ?? uuidv4();
  const room = rooms.get(roomId);

  if (room) {
    logger.info(`Room ${roomId} already exists.`);
    return room;
  }

  const server = new WebSocketServer({
    noServer: true,
  });

  const newRoom = {
    roomId,
    server,
    players: new Map<WebSocket, Player>(),
    turns: <PlaceFigureMessage[]>[],
  };
  rooms.set(roomId, newRoom);

  return newRoom;
};

export const getClients = (id: string): Set<WebSocket> =>
  rooms.get(id)?.server.clients;

export const getPlayers = (id: string): Map<WebSocket, Player> | null =>
  rooms.get(id)?.players || null;

export const getTurns = (id: string): Array<PlaceFigureMessage> =>
  rooms.get(id)?.turns;

export const printRooms = (): void => {
  for (const entry of rooms.entries()) {
    logger.debug(`${entry[0]} -> clients: ${entry[1].server.clients.size}`);
  }
};

export const cleanUp = (id: string): void => {
  if (getClients(id)?.size === 0) {
    logger.info(`Closing game room ${id}`);
    rooms.get(id)?.server.close();
    rooms.delete(id);
  }
};
