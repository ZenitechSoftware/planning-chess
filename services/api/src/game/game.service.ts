import { PlaceFigureMessage } from '../domain/messages';
import { calculateScore } from '../helpers/calculate-score';
import * as gameRoomService from './game-room.service';
import logger from '../logger';

export const playerHasMove = (roomId: string, playerId: string): boolean =>
  Boolean(findMoveByPlayerId(roomId, playerId));

export const playerHasSkipped = (roomId: string, playerId: string): boolean =>
  Boolean(gameRoomService.getSkippedPlayers(roomId).find(player => player.userId === playerId));

export const figureMoved = (
  roomId: string,
  payload: PlaceFigureMessage,
): PlaceFigureMessage[] => {
  let score;
  if (payload !== null) {
    score = calculateScore(payload);
  }

  if (playerHasMove(roomId, payload.id)) {
    removeTurn(roomId, payload.id);
  }

  gameRoomService.getTurns(roomId).push({ ...payload, score });
  return gameRoomService.getTurns(roomId);
};

export const clearBoard = (roomId: string): void => {
  gameRoomService.getTurns(roomId).length = 0;
};

export const findMoveByPlayerId = (
  roomId: string,
  id: string,
): PlaceFigureMessage | undefined =>
  gameRoomService.getTurns(roomId).find((turn) => turn.id === id);

export const getBoard = (roomId: string): PlaceFigureMessage[] =>
  gameRoomService.getTurns(roomId);

export const removeTurn = (roomId: string, playerId: string): void => {
  const turns = gameRoomService.getTurns(roomId);
  try {
    const turnIndex = turns.findIndex((turn) => turn.id === playerId);
    if (turnIndex < 0) {
      throw new Error(`Player with ${playerId} id move could not be found`);
    }
    turns.splice(turnIndex, 1);
  } catch (err) {
    logger.error(err?.message);
  }
};
