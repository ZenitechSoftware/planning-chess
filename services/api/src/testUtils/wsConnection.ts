import WebSocket from 'ws';
import { GameWebSocket } from '../domain/GameRoom';

const roomId = 'abcd-1234';

export const ws: GameWebSocket = new WebSocket('') as GameWebSocket;
ws.roomId = roomId;
