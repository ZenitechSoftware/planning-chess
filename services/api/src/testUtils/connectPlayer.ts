import * as playerService from '../messaging/players.service';
import {
  ReceivedMessage,
  MessageType,
  PlayerConnectedMessage,
} from '../domain/messages';
import { PlayerRole } from '../domain';
import { Player } from '../domain/player';
import { ws } from './wsConnection';

jest.mock('ws');

const playerTestId = 'some-short-v4-uuid-1';
const spectatorTestId = 'some-short-v4-uuid-2';
const roomId = 'abcd-1234';

export const voterConnect = (
  playerConfig: Partial<PlayerConnectedMessage> = {},
): Player => {
  const payload = {
    playerName: playerConfig.playerName ?? 'player1',
    id: playerConfig.id ?? playerTestId,
    role: playerConfig.role ?? PlayerRole.Voter,
    avatar: playerConfig.avatar ?? undefined,
  };

  const message: ReceivedMessage<MessageType.PlayerConnected> = {
    type: MessageType.PlayerConnected,
    payload,
  };

  playerService.newMessageReceived(ws, message);
  const [, voterPlayer] = playerService.findPlayerById(roomId, payload.id);
  return voterPlayer;
};

export const spectatorConnect = (): void => {
  const message: ReceivedMessage<MessageType.PlayerConnected> = {
    type: MessageType.PlayerConnected,
    payload: {
      playerName: 'spectator1',
      id: spectatorTestId,
      role: PlayerRole.Spectator,
      avatar: undefined,
    },
  };
  playerService.newMessageReceived(ws, message);
};
