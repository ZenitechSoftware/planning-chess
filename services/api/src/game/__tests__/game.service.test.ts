import WebSocket from 'ws';
import * as playerService from '../../messaging/players.service';
import * as gameService from '../game.service';
import * as gameRoomService from '../game-room.service';
import {
  MessageType,
  PlaceFigureMessage,
  ReceivedMessage,
} from '../../domain/messages';
import logger from '../../logger';
import { voterConnect } from '../../testUtils/connectPlayer';
import { TurnType } from '../../domain/game';
import { ws } from '../../testUtils/wsConnection';

jest.mock('ws');
jest.mock('../../logger');

describe('game.service', () => {
  const roomId = 'abcd-1234';
  const playerTestId = 'some-short-v4-uuid-1';
  const testTurn: PlaceFigureMessage = {
    row: 1,
    tile: 1,
    figure: 'rock',
    player: 'player1',
    id: playerTestId,
    score: 1,
  };
  const anotherTestTurn: PlaceFigureMessage = {
    row: 4,
    tile: 4,
    figure: 'rock',
    player: 'player1',
    id: playerTestId,
    score: 1,
  };

  const figureMoveMessage = (
    turn: PlaceFigureMessage,
  ): ReceivedMessage<MessageType.FigureMoved> => {
    return {
      type: MessageType.FigureMoved,
      payload: turn,
    };
  };

  beforeAll(() => {
    gameRoomService.getOrCreateRoom(roomId);
    jest.spyOn(global.Math, 'random').mockReturnValue(1);
    Object.defineProperty(ws, 'readyState', { value: WebSocket.OPEN });
  });

  afterEach(() => {
    jest.clearAllMocks();
    playerService.resetGame(ws);
    playerService.unsubscribe(ws);
  });

  it('should return false, when player has not moved, after the move should return true', () => {
    const connectedVoter = voterConnect();
    expect(
      gameService.playerHasPlacedFigure(roomId, connectedVoter.id),
    ).toBeFalsy();
    playerService.newMessageReceived(ws, figureMoveMessage(testTurn));
    expect(
      gameService.playerHasPlacedFigure(roomId, connectedVoter.id),
    ).toBeTruthy();
  });

  it('should return all turns', () => {
    const figureMovedSpy = jest.spyOn(gameService, 'figureMoved');
    voterConnect();
    const turnsCount = gameRoomService.getTurns(roomId).length;
    expect(turnsCount).toBe(0);
    playerService.newMessageReceived(ws, figureMoveMessage(testTurn));
    const turnsAfterMoveCount = gameRoomService.getTurns(roomId).length;
    expect(turnsAfterMoveCount).toBe(1);
    expect(figureMovedSpy).toReturnWith([
      { ...testTurn, turnType: TurnType.FigurePlaced },
    ]);
  });

  it('should return a turn by player id', () => {
    voterConnect();
    playerService.newMessageReceived(ws, figureMoveMessage(testTurn));
    const playerTurn = gameService.findMoveByPlayerId(roomId, playerTestId);
    expect(playerTurn).toEqual({
      ...testTurn,
      turnType: TurnType.FigurePlaced,
    });
  });

  it('should remove players turn', () => {
    voterConnect();
    playerService.newMessageReceived(ws, figureMoveMessage(testTurn));
    const playerTurn = gameService.findMoveByPlayerId(roomId, playerTestId);
    expect(playerTurn).toEqual({
      ...testTurn,
      turnType: TurnType.FigurePlaced,
    });
    const turnCount = gameRoomService.getTurns(roomId).length;
    expect(turnCount).toBe(1);

    gameService.removeTurn(roomId, playerTestId);
    const playerTurnAfter = gameService.findMoveByPlayerId(
      roomId,
      playerTestId,
    );
    expect(playerTurnAfter).toBeUndefined();
    const turnCountAfter = gameRoomService.getTurns(roomId).length;
    expect(turnCountAfter).toBe(0);
  });

  it('should remove players turn before marking player skipped move', () => {
    voterConnect();
    playerService.newMessageReceived(ws, figureMoveMessage(testTurn));
    const playerTurn = gameService.findMoveByPlayerId(roomId, playerTestId);
    expect(playerTurn).toEqual({
      ...testTurn,
      turnType: TurnType.FigurePlaced,
    });
    const turnCount = gameRoomService.getTurns(roomId).length;
    expect(turnCount).toBe(1);

    gameService.moveSkipped(roomId, playerTestId);
    const playerTurnAfterSkip = gameService.findMoveByPlayerId(
      roomId,
      playerTestId,
    );
    expect(playerTurnAfterSkip).toEqual({
      id: playerTestId,
      turnType: TurnType.MoveSkipped,
    });
    const turnCountAfterSkip = gameRoomService.getTurns(roomId).length;
    expect(turnCountAfterSkip).toBe(1);
  });

  it('should return true if player has skipped the move', () => {
    voterConnect();
    gameService.moveSkipped(roomId, playerTestId);
    const playerTurnAfterSkip = gameService.findMoveByPlayerId(
      roomId,
      playerTestId,
    );
    expect(playerTurnAfterSkip).toEqual({
      id: playerTestId,
      turnType: TurnType.MoveSkipped,
    });
    const hasPlayerSkipped = gameService.playerHasSkipped(roomId, playerTestId);
    expect(hasPlayerSkipped).toBe(true);
  });

  it('should not remove players turn, because players turn does not exists', () => {
    const connectedPlayer = voterConnect({
      id: playerTestId,
    });
    const playerTurn = gameService.findMoveByPlayerId(
      roomId,
      connectedPlayer.id,
    );
    expect(playerTurn).toBeUndefined();
    expect(() => gameService.removeTurn(roomId, playerTestId)).not.toThrow();
    expect(logger.error).toHaveBeenCalled();
  });

  it('should delete a player move, before assigning another move value', () => {
    const turnsCount = gameRoomService.getTurns(roomId).length;
    expect(turnsCount).toBe(0);
    voterConnect();
    playerService.newMessageReceived(ws, figureMoveMessage(testTurn));
    const turnsCountAfterFirstMove = gameRoomService.getTurns(roomId).length;
    expect(turnsCountAfterFirstMove).toBe(1);

    playerService.newMessageReceived(ws, figureMoveMessage(anotherTestTurn));
    const turnsCountAfterSecondMove = gameRoomService.getTurns(roomId).length;
    expect(turnsCountAfterSecondMove).toBe(1);
  });

  it('should clear the board, setting turns to empty array', () => {
    voterConnect();
    playerService.newMessageReceived(ws, figureMoveMessage(testTurn));
    const turnsCount = gameRoomService.getTurns(roomId).length;
    expect(turnsCount).toBe(1);
    const resetGameMessage: ReceivedMessage<MessageType.ClearBoard> = {
      type: MessageType.ClearBoard,
    };
    playerService.newMessageReceived(ws, resetGameMessage);
    const turnsCountAfterReset = gameRoomService.getTurns(roomId).length;
    expect(turnsCountAfterReset).toBe(0);
  });
});
