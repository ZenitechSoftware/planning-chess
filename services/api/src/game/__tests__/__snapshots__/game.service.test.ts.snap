// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game.service should return empty list when all players are done 1`] = `Array []`;

exports[`game.service should return new game state 1`] = `
Array [
  Object {},
]
`;
